migrator/tests/extract/test_kafka_extract.py

```diff
@@ -0,0 +1,260 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2023 CERN.
+#
+# ZenodoRDM is free software; you can redistribute it and/or modify it
+# under the terms of the MIT License; see LICENSE file for more details.
+
+"""Migrator tests configuration."""
+
+import itertools
+import random
+from collections import Counter
+from unittest.mock import PropertyMock
+
+from invenio_rdm_migrator.extract import Tx
+
+from zenodo_rdm_migrator.extract import KafkaExtract, KafkaExtractEnd
+
+
+def _random_chunks(li, min_chunk=1, max_chunk=50):
+    it = iter(li)
+    while True:
+        nxt = list(itertools.islice(it, random.randint(min_chunk, max_chunk)))
+        if nxt:
+            yield nxt
+        else:
+            break
+
+
+class MockConsumer(list):
+    """Mock Kafka consumer iterator."""
+
+    def commit(self):
+        pass
+
+
+def _patch_consumers(mocker, tx_info, ops):
+    """Helper for patching consumers of ``KafkaExtract``."""
+    mocker.patch.object(
+        KafkaExtract,
+        "_tx_consumer",
+        side_effect=[*tx_info, MockConsumer([]), KafkaExtractEnd],
+        new_callable=PropertyMock,
+    )
+    mocker.patch.object(
+        KafkaExtract,
+        "_ops_consumer",
+        side_effect=[*ops, MockConsumer([]), KafkaExtractEnd],
+        new_callable=PropertyMock,
+    )
+
+
+FIRST_TX_OP_COUNTS = (
+    563388798,
+    {
+        "public.oauth2server_token": 1,
+        "public.records_metadata": 1,
+    },
+)
+LAST_TX_OP_COUNTS = (
+    563390849,
+    {
+        "public.pidstore_pid": 8,
+        "public.files_bucket": 5,
+        "public.records_metadata": 3,
+        "public.files_files": 2,
+        "public.files_object": 2,
+        "public.records_buckets": 1,
+        "public.communities_community_record": 1,
+        "public.pidstore_redirect": 1,
+        "public.pidrelations_pidrelation": 1,
+    },
+)
+MIDDLE_TX_OP_COUNTS = (
+    563390343,
+    {
+        "public.files_bucket": 2,
+        "public.files_files": 2,
+        "public.files_object": 2,
+        "public.oauth2server_token": 1,
+        "public.records_metadata": 1,
+    },
+)
+
+
+def _assert_op_counts(tx, expected_counts):
+    op_counts = Counter(
+        [f'{op["source"]["schema"]}.{op["source"]["table"]}' for op in tx.operations]
+    )
+    assert op_counts == expected_counts
+
+
+def _assert_result(
+    result,
+    *,
+    count,
+    first_tx_id,
+    last_tx_id,
+    extra_tx_ids=None,
+    excluded_tx_ids=None,
+    tx_op_counts=None,
+):
+    assert len(result) == count
+    assert all(isinstance(t, Tx) for t in result)
+    tx_dict = {t.id: t for t in result}
+    tx_ids = list(tx_dict.keys())
+    assert tx_ids[0] == first_tx_id
+    assert tx_ids[-1] == last_tx_id
+    assert set(extra_tx_ids or []) <= set(tx_ids)
+    assert len(set(excluded_tx_ids or []).intersection(set(tx_ids))) == 0
```

Comment:
sets do not guarantee order, does this always pass?

Reply:
In this case, it's only testing exclusion of certain transactions, and not the order. 

Overall order is guaranteed though, because of the core logic of how the `KafkaExtract` yields transactions: the `self.tx_registry` is a `sortedcontainers.SortedDict`, which ensures insertion order of keys and yielded items are "popped" from the beginning. 

In any case, I should add a "sanity check" though of the results order :)